File: BSharp/Entities/ReportDefinition.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace BSharp.Entities
{
    [StrongEntity]
    public class ReportDefinitionForSave<TParameter,TRow, TColumn, TMeasure, TSelect> : EntityWithKey<string>
    {
        [MultilingualDisplay(Name = "Title", Language = Language.Primary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Title { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Secondary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Title2 { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Ternary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Title3 { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Primary)]
        [StringLength(1024, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Description { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Secondary)]
        [StringLength(1024, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Description2 { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Ternary)]
        [StringLength(1024, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Description3 { get; set; }

        [Display(Name = "ReportDefinition_Type")]
        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [AlwaysAccessible]
        [ChoiceList(new object[] { "Summary", "Details" }, 
            new string[] { "ReportDefinition_Type_Summary", "ReportDefinition_Type_Details" })]
        public string Type { get; set; }

        [Display(Name = "ReportDefinition_Chart")]
        [AlwaysAccessible]
        [ChoiceList(new object[] { "Card", "BarsVertical", "BarsHorizontal", "Line", "Pie" }, // TODO Add the rest, here and in TypeScript
            new string[] { "ReportDefinition_Chart_Card", "ReportDefinition_Chart_BarsVertical", "ReportDefinition_Chart_BarsHorizontal", "ReportDefinition_Chart_Line", "ReportDefinition_Chart_Pie" })]
        public string Chart { get; set; }

        [Display(Name = "ReportDefinition_DefaultsToChart")]
        [AlwaysAccessible]
        public bool? DefaultsToChart { get; set; }

        [Display(Name = "ReportDefinition_Collection")]
        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Collection { get; set; }

        [Display(Name = "ReportDefinition_DefinitionId")]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string DefinitionId { get; set; }

        [Display(Name = "ReportDefinition_Filter")]
        [StringLength(1024, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Filter { get; set; } // On drill down for summary

        [Display(Name = "ReportDefinition_OrderBy")]
        [StringLength(1024, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string OrderBy { get; set; } // On drill down for summary

        [Display(Name = "ReportDefinition_Top")]
        [AlwaysAccessible]
        public int? Top { get; set; }

        [Display(Name = "ReportDefinition_ShowColumnsTotal")]
        [AlwaysAccessible]
        public bool? ShowColumnsTotal { get; set; }

        [Display(Name = "ReportDefinition_ShowRowsTotal")]
        [AlwaysAccessible]
        public bool? ShowRowsTotal { get; set; }

        [Display(Name = "MainMenuSection")]
        [ChoiceList(new object[] { "Financials", "Administration" }, // TODO Add the rest, here and in TypeScript
            new string[] { "Financials", "Administration" })]
        [AlwaysAccessible]
        public string MainMenuSection { get; set; }

        [Display(Name = "MainMenuIcon")]
        [ChoiceList(new object[] { "clipboard", "chart-pie" }, // TODO Add the rest, here and in TypeScript
            new string[] { "Icon_Clipboard", "Icon_ChartPie" })]
        [AlwaysAccessible]
        public string MainMenuIcon { get; set; }

        [Display(Name = "MainMenuSortKey")]
        [AlwaysAccessible]
        public decimal? MainMenuSortKey { get; set; }

        [Display(Name = "Definition_State")]
        [ChoiceList(new object[] { "Draft", "Deployed", "Archived" },
            new string[] { "Definition_State_Draft", "Definition_State_Deployed", "Definition_State_Archived" })]
        [AlwaysAccessible]
        public string State { get; set; }

        [Display(Name = "ReportDefinition_Parameters")]
        [ForeignKey(nameof(ReportParameterDefinition.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TParameter> Parameters { get; set; }

        [Display(Name = "ReportDefinition_Rows")]
        [ForeignKey(nameof(ReportRowDefinition.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TRow> Rows { get; set; }

        [Display(Name = "ReportDefinition_Columns")]
        [ForeignKey(nameof(ReportColumnDefinition.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TColumn> Columns { get; set; }

        [Display(Name = "ReportDefinition_Measures")]
        [ForeignKey(nameof(ReportMeasureDefinition.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TMeasure> Measures { get; set; }

        [Display(Name = "ReportDefinition_Select")]
        [ForeignKey(nameof(ReportSelectDefinition.ReportDefinitionId))]
        [AlwaysAccessible]
        public List<TSelect> Select { get; set; }
    }

    public class ReportDefinitionForSave : ReportDefinitionForSave<ReportParameterDefinitionForSave, ReportRowDefinitionForSave, ReportColumnDefinitionForSave, ReportMeasureDefinitionForSave, ReportSelectDefinitionForSave>
    {

    }
    public class ReportDefinition : ReportDefinitionForSave<ReportParameterDefinition, ReportRowDefinition, ReportColumnDefinition, ReportMeasureDefinition, ReportSelectDefinition>
    {

    }

    public class ReportParameterDefinitionForSave : EntityWithKey<int>
    {
        public string ReportDefinitionId { get; set; }

        [Display(Name = "ReportDefinition_Key")]
        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Key { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Primary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Secondary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label2 { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Ternary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label3 { get; set; }

        [Display(Name = "ReportDefinition_IsRequired")]
        [AlwaysAccessible]
        public bool? IsRequired { get; set; }
    }

    public class ReportParameterDefinition : ReportParameterDefinitionForSave
    {

    }

    public class ReportSelectDefinitionForSave : EntityWithKey<int>
    {
        public string ReportDefinitionId { get; set; }

        [Display(Name = "ReportDefinition_Path")]
        [StringLength(1024, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Path { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Primary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Secondary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label2 { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Ternary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label3 { get; set; }
    }

    public class ReportSelectDefinition : ReportSelectDefinitionForSave
    {

    }
    public abstract class ReportDimensionDefinition : EntityWithKey<int>
    {
        public string ReportDefinitionId { get; set; }

        [Display(Name = "ReportDefinition_Path")]
        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [StringLength(1024, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Path { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Primary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Secondary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label2 { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Ternary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label3 { get; set; }

        [Display(Name = "ReportDefinition_OrderDirection")]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        [ChoiceList(new object[] { "asc", "desc" },
            new string[] { "ReportDefinition_OrderDirection_Asc", "ReportDefinition_OrderDirection_Desc" })]
        public string OrderDirection { get; set; }

        [Display(Name = "ReportDefinition_AutoExpand")]
        [AlwaysAccessible]
        public bool? AutoExpand { get; set; }
    }

    public class ReportColumnDefinitionForSave : ReportDimensionDefinition
    {

    }
    public class ReportColumnDefinition : ReportColumnDefinitionForSave
    {

    }
    public class ReportRowDefinitionForSave : ReportDimensionDefinition
    {

    }
    public class ReportRowDefinition : ReportRowDefinitionForSave
    {

    }

    public class ReportMeasureDefinitionForSave : EntityWithKey<int>
    {
        public string ReportDefinitionId { get; set; }

        [Display(Name = "ReportDefinition_Path")]
        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [StringLength(1024, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Path { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Primary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Secondary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label2 { get; set; }

        [MultilingualDisplay(Name = "Label", Language = Language.Ternary)]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        public string Label3 { get; set; }

        [Display(Name = "ReportDefinition_OrderDirection")]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [AlwaysAccessible]
        [ChoiceList(new object[] { "asc", "desc" },
            new string[] { "ReportDefinition_OrderDirection_asc", "ReportDefinition_OrderDirection_desc" })]
        public string OrderDirection { get; set; }

        [Display(Name = "ReportDefinition_Aggregation")]
        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [AlwaysAccessible]
        [ChoiceList(new object[] { "count", "sum", "avg", "max", "min" },
            new string[] {
                "ReportDefinition_Aggregation_count",
                "ReportDefinition_Aggregation_sum",
                "ReportDefinition_Aggregation_avg",
                "ReportDefinition_Aggregation_max",
                "ReportDefinition_Aggregation_min"
            })]
        public string Aggregation { get; set; }
    }

    public class ReportMeasureDefinition : ReportMeasureDefinitionForSave
    {

    }
}

File: BSharp/ClientApp/src/app/data/entities/report-definition.ts
// tslint:disable:variable-name
// tslint:disable:max-line-length
import { EntityForSave } from './base/entity-for-save';
import { TenantWorkspace } from '../workspace.service';
import { TranslateService } from '@ngx-translate/core';
import { SettingsForClient } from '../dto/settings-for-client';
import { EntityDescriptor } from './base/metadata';
import { EntityWithKey } from './base/entity-with-key';

export type ReportOrderDirection = 'asc' | 'desc';
export type ReportType = 'Summary' | 'Details';
export type Aggregation = 'count' | 'sum' | 'avg' | 'max' | 'min';
export type ChartType = 'Card' | 'BarsVertical' | 'BarsVerticalGrouped' | 'BarsVerticalStacked' |
    'BarsVerticalNormalized' | 'BarsHorizontal' | 'BarsHorizontalGrouped' | 'BarsHorizontalStacked' |
    'BarsHorizontalNormalized' | 'Line' | 'Area' | 'AreaStacked' | 'AreaNormalized' | 'Pie' | 'Doughnut' |
    'HeatMap' | 'TreeMap' | 'NumberCards' | 'Gauge' | 'Radar';
export type MainMenuSection = 'Financials' | 'Administration'; // TODO
export type MainMenuIcon = 'clipboard' | 'chart-pie';
export type DefinitionState = 'Draft' | 'Deployed' | 'Archived';

export class ReportDefinitionForSave<
    TParameter = ReportParameterDefinitionForSave,
    TRow = ReportRowDefinitionForSave,
    TColumn = ReportColumnDefinitionForSave,
    TMeasure = ReportMeasureDefinitionForSave,
    TSelect = ReportSelectDefinitionForSave> extends EntityForSave {

    Title: string;
    Title2?: string;
    Title3?: string;
    Description?: string;
    Description2?: string;
    Description3?: string;
    Type: ReportType; // summary or details
    Chart?: ChartType;
    DefaultsToChart: boolean; // ?
    Collection: string;
    DefinitionId?: string;
    Filter?: string;
    OrderBy?: string;
    Top?: number;
    ShowColumnsTotal: boolean;
    ShowRowsTotal: boolean;
    MainMenuSection: MainMenuSection;
    MainMenuIcon: MainMenuIcon;
    MainMenuSortKey: number;
    State: DefinitionState;

    Select: TSelect[];
    Parameters?: TParameter[];
    Rows: TRow[];
    Columns: TColumn[];
    Measures: TMeasure[];
}

export class ReportDefinition extends ReportDefinitionForSave<
    ReportParameterDefinition, ReportRowDefinition, ReportColumnDefinition, ReportMeasureDefinition, ReportSelectDefinition> {

}

export class ReportParameterDefinitionForSave extends EntityForSave {
    ReportDefinitionId: string | number;
    Key: string; // e.g. 'FromDate'
    Label?: string;
    Label2?: string;
    Label3?: string;
    IsRequired: boolean;
}

export class ReportParameterDefinition extends ReportParameterDefinitionForSave {

}

export class ReportSelectDefinitionForSave extends EntityForSave {
    ReportDefinitionId: string | number;
    Path: string;
    Label?: string;
    Label2?: string;
    Label3?: string;
}

export class ReportSelectDefinition extends ReportSelectDefinitionForSave {

}

export abstract class ReportDimensionDefinition extends EntityForSave {
    ReportDefinitionId: string | number;
    Path: string;
    Label?: string;
    Label2?: string;
    Label3?: string;
    OrderDirection?: ReportOrderDirection;
    AutoExpand: boolean;
}

export class ReportColumnDefinitionForSave extends ReportDimensionDefinition {

}

export class ReportColumnDefinition extends ReportColumnDefinitionForSave {

}

export class ReportRowDefinitionForSave extends ReportDimensionDefinition {

}

export class ReportRowDefinition extends ReportRowDefinitionForSave {

}

export class ReportMeasureDefinitionForSave extends EntityForSave {
    ReportDefinitionId: string | number;
    Path: string;
    Label?: string;
    Label2?: string;
    Label3?: string;
    OrderDirection?: ReportOrderDirection;
    Aggregation: Aggregation;
}

export class ReportMeasureDefinition extends ReportMeasureDefinitionForSave {

}

const _select = ['', '2', '3'].map(pf => 'Title' + pf);
let _settings: SettingsForClient;
let _cache: EntityDescriptor;

export function metadata_ReportDefinition(ws: TenantWorkspace, trx: TranslateService, _: string): EntityDescriptor {
    // Some global values affect the result, we check here if they have changed, otherwise we return the cached result
    if (ws.settings !== _settings) {
        _settings = ws.settings;
        _cache = {
            collection: 'ReportDefinition',
            titleSingular: () => trx.instant('ReportDefinition'),
            titlePlural: () => trx.instant('ReportDefinitions'),
            apiEndpoint: 'report-definitions',
            screenUrl: 'report-definitions',
            select: _select,
            orderby: ws.isSecondaryLanguage ? [_select[1], _select[0]] : ws.isTernaryLanguage ? [_select[2], _select[0]] : [_select[0]],
            format: (item: EntityWithKey) => (ws.getMultilingualValueImmediate(item, _select[0]) || trx.instant('Untitled')),
            properties: {
                Id: { control: 'text', label: () => trx.instant('Id') },
                Title: { control: 'text', label: () => trx.instant('Title') + ws.primaryPostfix },
                Title2: { control: 'text', label: () => trx.instant('Title') + ws.secondaryPostfix },
                Title3: { control: 'text', label: () => trx.instant('Title') + ws.ternaryPostfix },
                Description: { control: 'text', label: () => trx.instant('Description') + ws.primaryPostfix },
                Description2: { control: 'text', label: () => trx.instant('Description') + ws.secondaryPostfix },
                Description3: { control: 'text', label: () => trx.instant('Description') + ws.ternaryPostfix },
                Type: {
                    control: 'choice',
                    label: () => trx.instant('ReportDefinition_Type'),
                    choices: ['Summary', 'Details'],
                    format: (c: string) => trx.instant(`ReportDefinition_Type_${c}`)
                },
                Chart: {
                    control: 'choice',
                    label: () => trx.instant('ReportDefinition_Chart'),
                    choices: [ // Ordered by number of supported dimensions for ease of selection
                        // 0 Dimensions
                        'Card',
                        // 1 Dimension
                        'BarsVertical', 'BarsHorizontal', 'Pie', 'Doughnut', 'TreeMap', 'NumberCards', 'Gauge',
                        // 1 or 2 Dimensions
                        'Line', 'Area',
                        // 2 Dimensions
                        'BarsVerticalGrouped', 'BarsVerticalStacked', 'BarsVerticalNormalized', 'BarsHorizontalGrouped',
                        'BarsHorizontalStacked', 'BarsHorizontalNormalized', /* 'AreaStacked', 'AreaNormalized', 'Radar', */ 'HeatMap'],
                    format: (c: string) => trx.instant(`ReportDefinition_Chart_${c}`)
                },
                DefaultsToChart: { control: 'boolean', label: () => trx.instant('ReportDefinition_DefaultsToChart') },
                Collection: { control: 'text', label: () => trx.instant('ReportDefinition_Collection') },
                DefinitionId: { control: 'text', label: () => trx.instant('ReportDefinition_DefinitionId') },
                Filter: { control: 'text', label: () => trx.instant('ReportDefinition_Filter') },
                OrderBy: { control: 'text', label: () => trx.instant('ReportDefinition_OrderBy') },
                Top: { control: 'number', label: () => trx.instant('ReportDefinition_Top'), minDecimalPlaces: 0, maxDecimalPlaces: 0 },
                ShowColumnsTotal: { control: 'boolean', label: () => trx.instant('ReportDefinition_ShowColumnsTotal') },
                ShowRowsTotal: { control: 'boolean', label: () => trx.instant('ReportDefinition_ShowRowsTotal') },
                MainMenuSection: {
                    control: 'choice',
                    label: () => trx.instant('MainMenuSection'),
                    choices: ['Financials', 'Administration'],
                    format: (c: string) => trx.instant(c)
                },
                MainMenuIcon: {
                    control: 'choice',
                    label: () => trx.instant('MainMenuIcon'),
                    choices: ['clipboard', 'chart-pie'],
                    format: (c: string) => {
                        switch (c) {
                            case 'clipboard': return trx.instant('Icon_Clipboard');
                            case 'chart-pie': return trx.instant('Icon_ChartPie');
                            default: return c;
                        }
                    },
                },
                MainMenuSortKey: { control: 'number', label: () => trx.instant('MainMenuSortKey'), minDecimalPlaces: 2, maxDecimalPlaces: 2 },
                State: {
                    control: 'state',
                    label: () => trx.instant('Definition_State'),
                    choices: ['Draft', 'Deployed', 'Archived'],
                    format: (c: string) => trx.instant(`Definition_State_${c}`),
                    color: (c: string) => {
                        switch (c) {
                            case 'Draft': return '#6c757d';
                            case 'Deployed': return '#28a745';
                            case 'Archived': return '#dc3545';
                            default: return c;
                        }
                    }
                },
            }
        };

        if (!ws.settings.SecondaryLanguageId) {
            delete _cache.properties.Name2;
        }

        if (!ws.settings.TernaryLanguageId) {
            delete _cache.properties.Name3;
        }
    }

    return _cache;
}
